### PR TITLE
incremental-backup.md: Add documentation for new backup_mode attribute

### PR DESCRIPTION
**Pull request by shenitzky**
In order to fix https://bugzilla.redhat.com/1861674, a new 'backup_mode'
attribute should be added for each disk that participates in an incremental
backup. The 'backup_mode' attribute will indicate the type of backup that
was taken for each disk:

  - 'full': If a full backup was taken for the disk.
  - 'incremental': If an incremental backup was taken for the disk.

This new attribute allows the engine to take a mixed backup that includes
incremental for part of the disks and full backup for others. The mixed
backup is needed in case that the user includes disks that weren't part
of the given checkpoint ID (from_checkpoint_id), a full backup should be
taken for those disks while an incremental backup will be taken for the
other disks.

Change-Id: I57fa40a9c7087d97fcacfdcc8ff40b8f3bdf466c
Bug-Url: https://bugzilla.redhat.com/1861674
Signed-off-by: Eyal Shenitzky <eshenitz@redhat.com>

I confirm that this pull request was submitted according to the [contribution guidelines](https://github.com/oVirt/ovirt-site/blob/master/CONTRIBUTING.md): @shenitzky 

This pull request needs review by: @nirs

### CODE REVIEW

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 1]
```diff
@@ -284,7 +291,23 @@ Response:
...
+        ...
+    </disks>
+    <phase>initiailizing</phase>
+    <creation_date>
+</backup>
+```
+
+Response for an incremental backup which includes disks
+that wasn't part of the given `<from_checkpoint_uuid>`:
+```
```
Comment (sandrobonazzola) [inline source/develop/release-management/features/storage/incremental-backup.md:305]: please use ```xml

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 2]
```diff
@@ -640,14 +671,18 @@ backups tasks. Use during backup to track and montior backup.
...
+for every disk. The url will be used instead of the image path on the host when
+creating an image transfer for a disk. The backup mode is used to inform the user
+on the type of backup that was taken for the disk (`full`|`incremental`) so he will
+know what will be the requested context for downloading the backup disk extents
+using imageio API.
 
 - vm_backup_disk_map
   - backup_id: UUID
   - disk_id: UUID
   - backup_url: "nbd:unix:/tmp/<id>.sock:exportname=<sdb>" | "nbd://localhost:<12345>/<sdb>"
```
Comment (sandrobonazzola) [inline source/develop/release-management/features/storage/incremental-backup.md:684]: please keep value of backup_url key in `` or html will fail with invalid element <sbd> , <12345>, <id>

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 3]
```diff
@@ -232,6 +232,7 @@ Start full backup. The response phase indicates that the backup is `"initializin
 You need to poll the backup until the phase is `"ready"`. 
 Once the backup is ready the response will include `<to_checkpoint_id>` which
 should be used as the `<from_checkpoint_id>` in the next incremental backup.
+When starting a full backup, the `backup_mode` for all the disks will be `full` (`backup_mode="full"`)
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:235]: We should explain first that we have a backup_mode attribute. Maybe:

   Once the backup is ready, the response include the following elements:
   - to_checkpoint_id - what it means...
   - disks - what is means...

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 4]
```diff
@@ -266,6 +267,12 @@ You need to poll the backup until the phase is `"ready"`.
 Once the backup is ready the response will include `<to_checkpoint_id>` which
 should be used as the `<from_checkpoint_id>` in the next incremental backup.
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:269]: We should describe the normal flow first, when all disks have backup_mode="incremental".

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 5]
```diff
@@ -266,6 +267,12 @@ You need to poll the backup until the phase is `"ready"`.
...
 
+In case of starting an incremental backup which includes disks that weren't part
+of the given `<from_checkpoint_uuid>`, or new disks that were attached
+to the VM after the given `<from_checkpoint_uuid>` was taken, the system will detect
+those disks and a full backup will be taken for them (`backup_mode="full"`).
+The `backup_mode` for the rest of the the disks will be `incremental` (`backup_mode="incremental"`).
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:274]: This text is not clear, and does not explain how this API should be used.

I think we need something like this:

    Before downloading the data, the user must check if incremental
    backup is available for a particular disk. If the system can provide
    incremental backup data for a disk, the disk backup_mode will be
    "incremetnal". in this case the user can start an image transfer,
    and preform incremental backup.

    If the system cannot provide incremental backup data for a disk, the 
    disk backup_mode will be "full". The user can start an image
    transfer and perform full backup of this disk.  The next backup for 
    this disk can be incremental.

    The system cannot provide incremental backup when:
    - adding a new disk to a VM
    - adding existing disk to a backup, which was not included in a
      previous backup specified by the `<from_checkpoint_uuid>`
    - the system find that persistent dirty bitmaps are missing on
      storage

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 6]
```diff
@@ -496,6 +519,14 @@ missing in the image, or unknown bitmap exist, all the checkpoints on
 the disk and engine database must be deleted, and the current backup
 must be a full backup.
 
+### Setting the disks backup_mode before starting an incremental backup
+
+Before starting an incremental backup, the system will validate that all
+the disks that participate in the incremental backup already backed-up in
+in the given checkpoint ID. For disks which are part of the given checkpoint ID
+the backup mode will be `incremental`, if the disks are not part of the given
+checkpoint, the backup mode will be `full`.
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:528]: This should also work for the case when we cannot syncronize checkpoints with libvirt. I think currently we fail the backup. Instead, all disks should be marked as backup_mode="full". This eliminates the error for this case.

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 7]
```diff
@@ -640,14 +671,18 @@ backups tasks. Use during backup to track and montior backup.
...
 
-Add vm_backup_disk_map table. This table keeps the backup url for every
-disk. This url will be used instead of the image path on the host when
-creating an image transfer for a disk.
+Add vm_backup_disk_map table. This table keeps the backup url and the backup mode
+for every disk. The url will be used instead of the image path on the host when
+creating an image transfer for a disk. The backup mode is used to inform the user
+on the type of backup that was taken for the disk (`full`|`incremental`) so he will
+know what will be the requested context for downloading the backup disk extents
+using imageio API.
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:679]: The backup mode should also be used for the ticket "dirty" flag. ticket "dirty" should be set to true only if backup_mode is "incremental".

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 8]
```diff
@@ -766,10 +801,6 @@ socket:
   delete the corrupted bitmaps manually. This will fail HA VM flows and
   storage operations. (libvirt/qemu).
 
-- When creating a snapshot, an active bitmap should be copied to the new
-  top layer, and qemu should continue to track changes, writing into the
-  copied bitmap (libvirt).
-
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:772]: This change is not related, please send another patch for this.

File: source/develop/release-management/features/storage/incremental-backup.md [hunk 9]
```diff
@@ -781,24 +812,9 @@ socket:
...
 - Attaching disk with bitmaps to older qemu version will invalidate
   bitmaps, creating corrupted backups. how can we prevent this? (qemu).
 
-- Are bitmaps are copied down from the active layer during live merge
-  (libvirt/qemu) and cold merge (qemu-img)?
-
-- Need libvirt API to list all bitmaps in an image in a running VM
-  (libvirt).
-
```
Comment (nirs) [inline source/develop/release-management/features/storage/incremental-backup.md:801]: Same, don't mix unrelated changes with the API change.

**Review by nirs — CHANGES_REQUESTED** [review]
The design looks good, suggested how to make the text more clear.

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 5 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md:283]: I've added a new section for 'Disk backup_mode' before the explanation about imageio requests

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 1 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md]: I will add the tag in a different patch for the entire file.

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 2 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md]: Not related to this patch, I will fix it in a different patch

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 4 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md:278]: The normal flow starts with full and then incremental, we cannot jump to incremental.
 Also, I've changed a little bit the explanation order so it should  fine now.

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 6 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md:544]: added

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 7 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md]: Mentioned in the section that dedicated to backup_mode

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 9 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md:801]: restored all the unrelated notes, I will remove them in a separate patch

File: source/develop/release-management/features/storage/incremental-backup.md (hunk 8 above)
Comment (shenitzky) [reply source/develop/release-management/features/storage/incremental-backup.md:772]: ack
